### PR TITLE
Xeno Companion Crate

### PR DESCRIPTION
**Pull request by fenndragon**
# Description

Added a Xeno companion crate with an accompanying event as well as properly tweaked and readded specific non anomalous ventcritters events with major tweaks to their behavior and settings.

---

# Changelog

:cl:
- add: Added Xeno companion crate and accompanying event
- fix: Fixed Specific ventcritters events and re-enabled them

### CODE REVIEW

**Review by Memeji — APPROVED** [review]
I worked with Fenn on this a lot.
From our testing it seemed good, content-wise.

Balancing is all that may be needed, but we'll see about that after it is implemented.
Others will give their approval and request changes if needed.

**Comment by fenndragon** [comment]
commented out the ventcritters events until whatever is going wrong with the events can be fixed (the event manager firing events repeatedly for some reason/spamming events) as discussed with Fansana

**Review by Fansana — APPROVED** [review]

**Comment by Fansana** [comment]
Merge this during the weekend so we got time to fix bugs that may arise.

**Review by FoxxoTrystan — CHANGES_REQUESTED** [review]
Please use the "Floof" folder direction, or this is going to be a conflict merge nightmare.

**Comment by fenndragon** [comment]
> Please use the "Floof" folder direction, or this is going to be a conflict merge nightmare.

I was told not to do this on things that will likely end up going upstream, as some of the EE people have showed interest in my other work like this with the events and such this was going to probably end up the same way, which would mean pulling it back out of floof

File: Resources/Prototypes/Entities/Mobs/NPCs/xenopet.yml [hunk 1]
```diff
@@ -283,61 +347,62 @@
...
+  - type: InteractionPopup
+    successChance: 0.5
+    interactSuccessString: petting-success-reptile
+    interactFailureString: petting-failure-generic
+    interactSuccessSpawn: EffectHearts
+    interactSuccessSound:
+      path: /Audio/Animals/lizard_happy.ogg
+
 #
 #- type: entity
```
Comment (Mnemotechnician) [inline Resources/Prototypes/Entities/Mobs/NPCs/xenopet.yml:407]: Please make sure to move all of these events out of these files, at least when you backport this to EE. 

All events must be placed in events.yml.

File: Resources/Prototypes/Entities/Mobs/NPCs/xenopet.yml [hunk 2]
```diff
@@ -515,21 +474,21 @@
...
+        - state: green
+        - sprite: Mobs/Aliens/Xenos/drone.rsi
+          state: sleeping
+    - type: RandomSpawner
+      prototypes:
+        - MobXenoNeutralRouny
+        - MobXenoNeutralPraetorian
+        - MobXenoNeutralDrone
+        - MobXenoNeutralRavager
+      chance: 1
```
Comment (Mnemotechnician) [inline Resources/Prototypes/Entities/Mobs/NPCs/xenopet.yml:494]: Not sure, but this probably belongs in a file with other mob spawners, like renault and other pets. This should be moved.

File: Resources/Prototypes/Entities/Objects/Specific/rehydrateable.yml [hunk 3]
```diff
@@ -266,3 +266,263 @@
         hard: false
         layer:
         - LowImpassable
+
```
Comment (Mnemotechnician) [inline Resources/Prototypes/Entities/Objects/Specific/rehydrateable.yml:269]: All of these should share a common parent to reduce verbosity. Introduce a base parent like BaseDehydratedXeno, move all common components from these items to that base, and make those items inherit from the new base item (you can make a prototype inherit from multiple parents by supplying a list to the `parent` property, e.g.: `parent: [BaseDehydratedXeno, PlushieFRouny]`. Do note that components from the first parents are inherited first, so if multiple parents have the same component, the resulting prototype will inherit it from the first parent only).

You could also do the same to other rehydratables - I don't know why wizden didn't introduce a common parent for them.

File: Resources/Prototypes/Entities/Mobs/NPCs/xenopet.yml [hunk 4]
```diff
@@ -23,82 +28,141 @@
...
-#    rules: ghost-role-information-friendlyxeno-rules
-#  - type: GhostTakeoverAvailable
-#  - type: LanguageKnowledge
-#    speaks:
-#    - Xeno
-#    understands:
-#    - Xeno
-#    - TauCetiBasic
-#
+
```
Comment (Mnemotechnician) [inline Resources/Prototypes/Entities/Mobs/NPCs/xenopet.yml:31]: These neutral xenos should also preferably share a common parent like MobXenoNeutralBase. Again, you can use multiple inheritance.

**Review by Mnemotechnician — CHANGES_REQUESTED** [review]
Please, avoid simply copy-pasting prototypes if you are only going to change a few components in them; use inheritance instead.

**Comment by fenndragon** [comment]
> Please, avoid simply copy-pasting prototypes if you are only going to change a few components in them; use inheritance instead.

I carefully considered what to parent and what not. In this instance the differences between the things resulted in parenting being less useful. 

I parent the neutral versions from the base version. With changes applied where acceptable, then the specific versions from those. 

To change the way this gets parented would result in more code not less. As I would need to respecify all changes listed in each specific xeno base already.

**Comment by FoxxoTrystan** [comment]
> > Please use the "Floof" folder direction, or this is going to be a conflict merge nightmare.
> 
> I was told not to do this on things that will likely end up going upstream, as some of the EE people have showed interest in my other work like this with the events and such this was going to probably end up the same way, which would mean pulling it back out of floof

Please still do this until its is on EE.
We will handle the conflict merge issues when EE has it and we do an upstream merge but for the moment this is requried here if not on upstream already.

**Comment by Mnemotechnician** [comment]
> I carefully considered what to parent and what not. In this instance the differences between the things resulted in parenting being less useful.
> 
> I parent the neutral versions from the base version. With changes applied where acceptable, then the specific versions from those.
> 
> To change the way this gets parented would result in more code not less. As I would need to respecify all changes listed in each specific xeno base already.

I really don't see how

<details>

```yml
- type: entity
  name: Friendly NT Ravager
  id: MobXenoNeutralRavager
  parent: MobXenoRavagerNPC
  description: A friendly ravager programmed by centcomm
  components:
  - type: NpcFactionMember
    factions:
    - PetsNT
  - type: LayingDown
  - type: Sprite
    drawdepth: Mobs
    sprite: Mobs/Aliens/FXS/ravager.rsi
    scale: 0.7, 0.7
  - type: PointLight
    radius: 2
    energy: 1
    color: "#E3954D"
  - type: GhostRole
    name: ghost-role-information-friendlyxeno-name
    description: ghost-role-information-friendlyxeno-description
    rules: ghost-role-information-friendlyxeno-rules
  - type: GhostTakeoverAvailable
  - type: LanguageKnowledge
    speaks:
    - Xeno
    understands:
    - Xeno
    - TauCetiBasic
```

</details>

Repeated 4-5 times is any shorter than

<details>

```yml
- type: entity
  name: Friendly NT Ravager
  id: MobXenoNeutralBase
  abstract: true
  components:
  - type: NpcFactionMember
    factions:
    - PetsNT
  - type: LayingDown
  - type: Sprite
    drawdepth: Mobs
    scale: 0.7, 0.7
  - type: PointLight
    radius: 2
    energy: 1
  - type: GhostTakeoverAvailable
  - type: LanguageKnowledge
    speaks:
    - Xeno
    understands:
    - Xeno
    - TauCetiBasic
 ```

Followed by this, repeated 4-5 times

```yml
- type: entity
  name: Friendly NT Ravager
  id: MobXenoNeutralRavager
  parent: [MobXenoNeutralBase, MobXenoRavager]
  description: A friendly ravager programmed by centcomm
  components:
  - type: Sprite
    sprite: Mobs/Aliens/FXS/ravager.rsi
  - type: PointLight
    color: "#E3954D"
```
</details>

**Comment by fenndragon** [comment]
> > I carefully considered what to parent and what not. In this instance the differences between the things resulted in parenting being less useful.
> > I parent the neutral versions from the base version. With changes applied where acceptable, then the specific versions from those.
> > To change the way this gets parented would result in more code not less. As I would need to respecify all changes listed in each specific xeno base already.
> 
> I really don't see how
> 
> ```yaml
> - type: entity
>   name: Friendly NT Ravager
>   id: MobXenoNeutralRavager
>   parent: MobXenoRavagerNPC
>   description: A friendly ravager programmed by centcomm
>   components:
>   - type: NpcFactionMember
>     factions:
>     - PetsNT
>   - type: LayingDown
>   - type: Sprite
>     drawdepth: Mobs
>     sprite: Mobs/Aliens/FXS/ravager.rsi
>     scale: 0.7, 0.7
>   - type: PointLight
>     radius: 2
>     energy: 1
>     color: "#E3954D"
>   - type: GhostRole
>     name: ghost-role-information-friendlyxeno-name
>     description: ghost-role-information-friendlyxeno-description
>     rules: ghost-role-information-friendlyxeno-rules
>   - type: GhostTakeoverAvailable
>   - type: LanguageKnowledge
>     speaks:
>     - Xeno
>     understands:
>     - Xeno
>     - TauCetiBasic
> ```
> 
> Repeated 4-5 times is any shorter than
> 
> ```yaml
> - type: entity
>   id: MobXenoNeutralBase
>   abstract: true
>   components:
>   - type: NpcFactionMember
>     factions:
>     - PetsNT
>   - type: LayingDown
>   - type: Sprite
>     drawdepth: Mobs
>     scale: 0.7, 0.7
>   - type: PointLight
>     radius: 2
>     energy: 1
>   - type: GhostTakeoverAvailable
>   - type: LanguageKnowledge
>     speaks:
>     - Xeno
>     understands:
>     - Xeno
>     - TauCetiBasic
> ```
> 
> Followed by this, repeated 4-5 times
> 
> ```yaml
> - type: entity
>   name: Friendly NT Ravager
>   id: MobXenoNeutralRavager
>   parent: [MobXenoNeutralBase, MobXenoRavagerNPC]
>   description: A friendly ravager programmed by centcomm
>   components:
>   - type: Sprite
>     sprite: Mobs/Aliens/FXS/ravager.rsi
>   - type: PointLight
>     color: "#E3954D"
> ```

(Parent:MobxenoravagerNPC) On that particular instance. Comparing size of the different parents that to the size of what I have as I would need to put everything from the parent at the end. I will gladly make an estimate of the file size in question though if it satisfies you.

**Comment by fenndragon** [comment]
told you

The file size of Xenopet in the example is merely the code of the file the other is attempting to replicate. 

I rest my case.

**Comment by Mnemotechnician** [comment]
It's about verbosity, readability and maintainability, not file size. If you are going to upstream this, you need to adhere to the (relatively low compared to most other projects) code quality standards we have in place. If you want to leave this as a floof-only thing, then it's fine, code quality hardly matters.

**Comment by fenndragon** [comment]
> It's about verbosity, readability and maintainability, not file size. If you are going to upstream this, you need to adhere to the (relatively low compared to most other projects) code quality standards we have in place. If you want to leave this as a floof-only thing, then it's fine, code quality hardly matters.

code is verbose and readable. each creature references its base creature as a parent and doesn't repeat any of the necessary code. the recolor/sprite changed critters are parented to their original base friendly version, I fail to understand how any of this is done incorrectly as every part of the code is necessary in my opinion. 

if you wish to have this recoded in a different manner, feel free to make the changes you wish to see and comment them, I'll gladly accept the changes if they are functional.

**Comment by Mnemotechnician** [comment]
I already suggested what could be done, see above in the &lt;details&gt; blocks. If you want some statistics, each prototype could be 64% shorter in terms of linecount and 71% shorter in terms of component count.

File: Resources/Prototypes/Entities/Objects/Specific/rehydrateable.yml (hunk 3 above)
Comment (fenndragon) [reply Resources/Prototypes/Entities/Objects/Specific/rehydrateable.yml:269]: the stuff in question has been moved to something in another folder, I don't want to change basegame files though as that's not my pull request.

**Review by Fansana — APPROVED** [review]

**Pull request merged**